Prediction tables in database: retrieve prediction from db, or generate and store in db if not there

cheerstopaula opened the pull request:
This PR consists in the addition of the prediction database framework, to be able to:
- generate the predictions in the form of hexagons and scores,
- store the hexagons indexes with their scores in database, 
- retrieve from database when generating prediction, and
- if prediction is not in database, generate hexagon indexes and scores and store in db, and finally
- allow changing the threshold in the sidebar. 

More specifically, this PR considers the following changes:

db/model.py
- prediction and prediction_hexagon: new tables in the database (see db/models.py)

tools.py
- generate_prediction_scores function: given eval_params, generates the prediction, computes the scores for all hexagons by taking the maximum value predicted within the hexagon (here, I considered a minimum score of 0.0001, otherwise we would be saving almost every hexagon with a non significant score), and returns hexagon indexes with their corresponding scores
- populate_prediction_database function: given eval_params, run generate_prediction_scores, and with the hexagon indexes and scores, populates the database according to the database schema
- get_predicted_hexagons: retrieves from database the hexagon indexes that comply with eval_params (taxa_id and threshold)

main.py
- generate_prediction: retrieves the hexagons by calling crud.py. If that taxa is not in the database, runs populate_prediction_database, to populate database with that species and returns the hexagons.

Sidebar.js:
- allows changing the threshold now, updates the prediction accordingly.

SergePogorelov commented on src/backend/app/tools.py:
```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
+def get_predicted_hexagons(db, eval_params):
+    taxa_name = eval_params["taxa_name"]
+    taxa_id = int(taxa_name.split("(")[-1][:-1])
+    try:
```
I would recommend using `try ... except` only if you really need it.
Now all errors are caught here, so we may not notice an error that is not related to the lack of prediction in the database

It's easier to check if prediction was found in the database, and than find hexagons, otherwise, return None

here is how I searching for the latest annotation  in db:
```
    latest_annotation = (
        db.query(models.Annotation)
        .filter(models.Annotation.taxa_id == taxa_id)
        .order_by(models.Annotation.created_at.desc())
        .first()
    )
```
you can do something like this, it looks more clear for me...

ginic replied:
This is a good point. It's good to make exceptions specific, so if there's a particular type of exception you know you want to handle here, just handle that and let all other types of exceptions get raised so that they don't fail silently.

cheerstopaula replied:
@SergePogorelov This was what I was trying to do but wasn't sure, used your solution and it's working properly. Thank you so much!
@ginic Noted! Will get rid of the other try except I have.

SergePogorelov commented on src/backend/app/tools.py:
```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
+    # if a more detailed HeatMap needed, use `pred_loc_combined` for that
+    scores_dict={}
+    for lat, lon, pred in pred_loc_combined:
+        try:
```
here too it is better to simply check if such a key is in the dictionary, and if not, then add it with an empty list, something like this:

```python
    scores_dict = {}
    for lat, lng, pred in pred_loc_combined:
        hex_index = h3.geo_to_h3(lat, lng, hex_resolution)
        if hex_index not in scores_dict:
            scores_dict[hex_index] = []
        scores_dict[hex_index].appnd(pred)
```

but if we only need the maximum score, then why just store it, no need to create a list for this

```python
    scores_dict = {}
    for lat, lng, pred in pred_loc_combined:
        hex_index = h3.geo_to_h3(lat, lng, hex_resolution)
        scores_dict[hex_index] = max(scores_dict.get(hex_index, 0), pred)
```

cheerstopaula replied:
This is what I mean when I say I code like a mathematician :D. Changed to first solution! I take your point, but I use the dictionary later on in other functions. Will consider changing it later.

SergePogorelov commented on src/backend/app/tools.py:
```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
+        hex_indexes.append(key)
+        hex_scores.append(max(scores_dict[key]))
+    
+    index_score_combined = np.column_stack((hex_indexes, hex_scores))
```
and here, if we already have only max scores in our ` scores_dict`, then we can do it like this
```python
return np.column_stack((scores_dict.keys(), scores_dict.values()))
```

but i think we can use `scores_dict` in dict format later. No need to convert it to np table.

cheerstopaula replied:
In this case, scores_dict has all the values in the form:
'872fffffff' : [0.123, 0.543, 0.345]
so scores_dict is storing all values, not the maximum one, therefore 
return np.column_stack((scores_dict.keys(), scores_dict.values()))
would not return the max to each hexagon index. 

I realize that this part of the code works fast and properly, but it's not really elegant. I'll try to simplify it.

SergePogorelov commented on src/backend/app/tools.py:
```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
+    pred_loc_combined = pred_loc_combined[pred_loc_combined[:, 2] >= lowest_threshold]
+    hex_resolution = eval_params['hex_resolution']
+
+    # if a more detailed HeatMap needed, use `pred_loc_combined` for that
```
this comment is no longer relevant)

SergePogorelov commented on src/backend/app/main.py:
```diff
@@ -43,9 +43,32 @@ async def health_check():
+    db = next(get_db())
+    predicted_hexagons=tools.get_predicted_hexagons(db, eval_params=eval_params)
+    if predicted_hexagons is None:
+        index_score_combined=tools.populate_prediction_database(eval_params,db)
```
if `populate_prediction_database` returns a dictionary, this part can be simplified

```python
predicted_hexagons = [
    kex_index for kex_index, kex_score in  index_score_combined.items() 
    if  float(kex_score) >= eval_params["threshold"]
]
```

cheerstopaula replied:
It does not return a dictionary, it returns:
 index_score_combined = np.column_stack((hex_indexes, hex_scores)). 
but I accept other suggestions on making 
predicted_hexagons = [indexes[i] for i in range(len(indexes)) if scores[i]>= eval_params["threshold"]]
better!

SergePogorelov commented on src/backend/app/main.py:
```diff
@@ -43,9 +43,32 @@ async def health_check():
+        "absence": list()
+    }
+    
+    if len(predicted_hexagons)==0:
```
I think this block is unnecessary here, because if `len(predicted_hexagons)==0`, then this is an empty list anyway

cheerstopaula replied:
Yes, I agree, however whenever I take that block, I get an error
<img width="1047" alt="Screen Shot 2024-08-06 at 1 47 35 PM" src="https://github.com/user-attachments/assets/23d3b371-dfc0-4492-8ce7-bef536d96903">
I'll go over it again and see if I can simplify it, not sure what the problem is.

cheerstopaula replied:
Simplified to:
    if len(predicted_hexagons)==0:
        response=dict(annotation_hexagon_ids=annotation_hexagon_ids)

    else:
        response = dict(
            prediction_hexagon_ids=predicted_hexagons,
            annotation_hexagon_ids=annotation_hexagon_ids,
        )

SergePogorelov commented on src/backend/app/tools.py:
```diff
@@ -1,10 +1,11 @@
 
-import alphashape
-from shapely.geometry import mapping
+# import alphashape
```
there is a lot of commented code in the repo related to `hull_points`. if we are not going to use them anymore, then I think we can completely delete all this code, so as not to confuse developers in the future, otherwise they will not understand why we left it.

cheerstopaula replied:
Didn't want to completely delete in case you preferred keeping the hull points in the future. Deleted now!

SergePogorelov commented on src/backend/app/tools.py:
```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
         {h3.geo_to_h3(lat, lon, hex_resolution) for lat, lon in coordinates}
```
It seems that `generate_prediction` function is not used anywhere else and can be removed.

SergePogorelov reviewed (changes requested):
Thanks Paula!
Realy great job!

I apologize for the large number of comments. 
I just like Python more than JavaScript, so I can give a more suggestions)

It is not necessary to fix all of them, but some of them could help simplify the code.

### The main things I would like to fix:
1. Remove commented code and unused functions.

2. Fix raff's suggestions for code style. Unfortunately, it checks each commit separately, and some of your commits did not pass the style.
The full list can be found at this link with red crosses.
https://github.com/UMassCDS/ds4cg2024-inaturalist/actions?query=branch%3Apaula%2Fprediction_database+event%3Apush

3. Update CHANGELOG

ginic commented on src/backend/app/db/models.py:
```diff
@@ -14,3 +14,15 @@ class AnnotationHexagon(Base):
+    __tablename__="prediction"
+    prediction_id = Column(Integer, primary_key=True, index=True)
+    taxa_id = Column(Integer)
+
```
I think we can ignore this for now. The database proposal has this linked via a foreign key to the `model` table with the `created_at` stored for the model. Since we haven't implemented that table yet, I think it's okay to leave out creation times for now.

cheerstopaula replied:
The created_at column is not in the MVB database design (see https://docs.google.com/document/d/1yryxqj19b_SocJbYjW3NjGAeexVmYe1-DRdXUD8ed5Q/edit), but it's nice to have, will add i.

cheerstopaula replied:
Added!

cheerstopaula commented on src/backend/app/main.py:
```diff
@@ -43,9 +43,32 @@ async def health_check():
-    response = tools.generate_prediction(await request.json())
-    return JSONResponse(content=response)
+    eval_params=await request.json()
+    db = next(get_db())
```
I tried a bunch of different things before everything worked. Changed it back to the original way of calling the db to be consistent with the other functions. Seems to be working properly

ginic reviewed (approved):
This looks great, @cheerstopaula ! I tested it out and checked that the database saves and loads as expected. Database schema looks good and matches the design. 

The one thing I noticed is that the prediction displays are not updated when you change the threshold unless you click "Generate Prediction" again. This is a little bit of clunkiness in the UI that might confuse users. I don't know if that would be quick to fix though. If it's easy to fix, you can add it here, but if it's not easy, you can merge this and we'll just make a bug about this to address in the future.

cheerstopaula commented:
> Thanks Paula! Realy great job!
> 
> I apologize for the large number of comments. I just like Python more than JavaScript, so I can give a more suggestions)
> 
> It is not necessary to fix all of them, but some of them could help simplify the code.
> 
> ### The main things I would like to fix:
> 1. Remove commented code and unused functions.
> 2. Fix raff's suggestions for code style. Unfortunately, it checks each commit separately, and some of your commits did not pass the style.
>    The full list can be found at this link with red crosses.
>    https://github.com/UMassCDS/ds4cg2024-inaturalist/actions?query=branch%3Apaula%2Fprediction_database+event%3Apush
> 3. Update CHANGELOG

Thanks @SergePogorelov for the great comments!! Specially the suggestions to avoid the try-excepts that I was unceremoniously using. 
1. Removed commented code and unused functions
2. Addressed the raff suggestions and passing tests now. 
3. Updated CHANGELOG accordingly

cheerstopaula commented:
> This looks great, @cheerstopaula ! I tested it out and checked that the database saves and loads as expected. Database schema looks good and matches the design.
> 
> The one thing I noticed is that the prediction displays are not updated when you change the threshold unless you click "Generate Prediction" again. This is a little bit of clunkiness in the UI that might confuse users. I don't know if that would be quick to fix though. If it's easy to fix, you can add it here, but if it's not easy, you can merge this and we'll just make a bug about this to address in the future.

Thanks @ginic ! I agree that it would be more intuitive, however, I think this is not a quick fix, I think @SergePogorelov was struggling with something similar? I can try to look into it.

oz03-hub commented on src/frontend/src/components/Sidebar.js:
```diff
@@ -112,10 +112,9 @@ const Sidebar = forwardRef((props, ref) => {
         defaultValue="0.1"
-        min="0.1"
-        max="0.9"
+        min="0.01"
```
why don't we set min="0.0" and max="1.0"

cheerstopaula replied:
Good question! So the thing is that a lot of hexagons had a score<e-23 (that's an example number, just to say a very very small number, a model is rarely going to give you zero), so each prediction was storing waaaay too many hexagons. So I put a minimum threshold of 0.0001 (arbitrarily) to filter them out. Thus, min could not be zero because it wouldn't make sense. Since I set the step to be 0.01, the min had to be 0.01. The max I arbitrarily set to 0.99 (considering the 0.01 step) instead of 1 though. If we set it to 1, it will probably just show an empty map (the model is rarely going to give you an exact 0 or 1). I hope that clears it out, feel free to change it too if you think something different is more appropriate!

oz03-hub replied:
the reasoning for min makes sense, I also noticed that species don't have much datapoints above 0.6-0.7 threshold, it isn't related to this ticket but something interesting

oz03-hub reviewed (approved):
Things are working I left a question because I am curios, otherwise things work as expected. Thanks for not renaming annotation table names my data extractor still works.

ginic commented:
> > This looks great, @cheerstopaula ! I tested it out and checked that the database saves and loads as expected. Database schema looks good and matches the design.
> > The one thing I noticed is that the prediction displays are not updated when you change the threshold unless you click "Generate Prediction" again. This is a little bit of clunkiness in the UI that might confuse users. I don't know if that would be quick to fix though. If it's easy to fix, you can add it here, but if it's not easy, you can merge this and we'll just make a bug about this to address in the future.
> 
> Thanks @ginic ! I agree that it would be more intuitive, however, I think this is not a quick fix, I think @SergePogorelov was struggling with something similar? I can try to look into it.

@cheerstopaula , no problem! Don't worry about fixing it here then, if you can gather more details that'd be awesome and we can report a bug together.

SergePogorelov reviewed (approved):
Great Job!
